lib/block-write-stream.ts

```diff
@@ -56,10 +62,50 @@ export class BlockWriteStream extends Writable {
 			return;
 		}
 
-		this.__write(buffer, this.bytesWritten).then(callback, callback);
+		if (this._bytes === 0 && buffer.length >= this.chunkSize) {
+			if (buffer.length % this.blockSize === 0) {
+				this.__write(buffer, this.bytesWritten).then(callback, callback);
+				return;
+			}
+		}
+
+		this._buffers.push(buffer);
+		this._bytes += buffer.length;
+
+		if (this._bytes >= this.chunkSize) {
+			let block = Buffer.concat(this._buffers);
+			const length = Math.floor(block.length / this.blockSize) * this.blockSize;
+
+			this._buffers.length = 0;
+			this._bytes = 0;
+
+			if (block.length !== length) {
+				this._buffers.push(block.slice(length));
+				this._bytes += block.length - length;
+				block = block.slice(0, length);
+			}
+
+			this.__write(block, this.bytesWritten).then(callback, callback);
```

Comment:
Why is the callback used both for resolution and rejection? Who sets the err argument to `null` in the success case?

Reply:
That's a very bad reason to do it this way. Use `Promise.asCallback` or pass `null` manually

http://bluebirdjs.com/docs/api/ascallback.html

Reply:
As this is a native `Promise`, there's no `.asCallback()` available, and the return type of an async function has to be a native Promise – so the best thing I can come up with, without turning everything else into using callbacks, would be

```ts
function asCallback(promise: Promise<any>, callback: (error: Error | void, value?: any) => void) {
	promise.then((value: any) => {
		callback(undefined, value);
	}).catch(callback)
}
```

wdyt?

Reply:
Or, of course we could inline the `.then((value) => callback(null, value))`

Reply:
Whatever you prefer. Yet another way is:

```javascript
Promise.resolve(this.__write(block, this.bytesWritten)).asCallback(callback)
```

where Promise is bluebird's Promise